# Prerequisites

On top of a thirst for knowledge, and a working Internet connection, here is what you will need for the course : 


## NGS

As announced in the [course registration webpage](https://www.sib.swiss/training/course/20220901_IRNAS),
we expect participants to already have a basic knowledge in Next Generation Sequencing (NGS) techniques. 


## UNIX

Practical knowledge of the UNIX command line is also required to be able to follow this course, given that that the tools used to process sequenced reads use this interface.

If you are unsure about your capabilities or feel a bit rusty, we strongly recommend you spend some time practicing before the course : in our experience, the more comfortable you are with UNIX, the more you will be able to focus on the RNA-seq during the course, and the more you will gain from it.

You may refer to the [SIB's UNIX e-learning module](https://edu.sib.swiss/pluginfile.php/2878/mod_resource/content/4/couselab-html/content.html),
or follow our small [UNIX crash course page](days/unix_crash.md)




## R 

A basic knowledge of the [R language](https://www.r-project.org/) is required to perform most analytical steps after reads have been mapped and quantified : differential gene expression, gene set enrichment, over-representation/pathway analysis.

If you are not familiar with R, you can follow our small [R crash course page](days/R_crash.md), but given the depth of the subject, we also recommend you attend a dedicated course, such as the [SIB First Steps with R course](https://github.com/sib-swiss/first-steps-with-R-training), or you can pick one from [glittr.org](https://glittr.org/)




<!-- Suggestion: should we have some practical notes re: the teaching location?
Coffee/snakcs/toilets...? Or should this be external to the course website? -->
